src/parsers/gene_association_file.py

```diff
@@ -0,0 +1,183 @@
+"""
+GO Gene Association File (GAF) parser.
+
+This script processes Gene Ontology (GO) annotation data from GAF-style CSV files and normalizes it into a structured format compatible with ontology-driven data models.
+
+Main Functionality:
+--------------------
+- Loads raw GO annotations from an input CSV file.
+- Drops unused or redundant columns to simplify the dataset.
+- Constructs a unified 'subject' identifier by combining 'DB' and 'DB_Object_ID'.
+- Renames selected columns to match a target schema (e.g., predicate, object, supporting_objects).
+- Converts annotation dates into standard YYYY-MM-DD format.
+- Adds metadata fields such as 'aggregator' and 'protocol_id'.
+- Determines predicate negation based on a whitelist of allowed GO relations.
+- Maps evidence codes to ECO terms by merging with an external mapping file.
+- Applies fallback evidence types when no direct evidence mapping is found.
+- Outputs a clean, standardized CSV file for downstream processing.
+
+Resources:
+----------
+- Gene Association File (GAF) format specification: https://geneontology.org/docs/go-annotation-file-gaf-format-2.2/
+- Downloadable GAF files: http://current.geneontology.org/annotations/index.html
+- ECO (Evidence and Conclusion Ontology) documentation: https://www.evidenceontology.org/annotation_resources/#eco_go_mappings
+- GAF-to-ECO mapping file: http://purl.obolibrary.org/obo/eco/gaf-eco-mapping.txt
+
+Usage:
+------
+Run via CLI:
+    python gene_association_file.py --input annotations_data100.csv --output normalized_annotation.csv
+
+"""
+
+import pandas as pd
+import click 
+
+# --- Load ECO mapping ---
+ECO_MAPPING_URL = "http://purl.obolibrary.org/obo/eco/gaf-eco-mapping.txt"
+mapping_columns = ["Evidence_Code", "DB_Reference", "evidence_type"]
+
+try:
+    eco_mapping_df = pd.read_csv(ECO_MAPPING_URL, sep="\t", comment="#", header=None, names=mapping_columns)
+except Exception as e:
+    raise ConnectionError(f"Failed to load ECO mapping from {ECO_MAPPING_URL}: {e}")
+
+def process_go_annotations(input_path: str, output_path: str):
+
+    """
+    Processes GO annotations by cleaning, transforming, and mapping evidence types.
+
+    Args:
+        input_path (str): Path to the input GO annotations CSV file.
+        output_path (str): Path to save the normalized output CSV file.
+    """
+    
+    # load the input data 
+    go_annotations_df = pd.read_csv(input_path)
+    
+    # Check if essential columns exist in the metadata 
+    requeired_columns = {"DB","DB_Object_ID","Qualifier","GO_ID",
+                         "DB_Reference","Evidence_Code","With_From",
+                         "Date","Assigned_By"}
+    
+    if not requeired_columns.issubset(go_annotations_df.columns):
+        missing_columns = requeired_columns - set(go_annotations_df.columns)
```

Comment:
```suggestion
    required_columns = {"DB","DB_Object_ID","Qualifier","GO_ID",
                         "DB_Reference","Evidence_Code","With_From",
                         "Date","Assigned_By"}
    
    if not required_columns.issubset(go_annotations_df.columns):
        missing_columns = required_columns - set(go_annotations_df.columns)
```

Typo in variable name :)